src/services/cache-service.ts

```diff
@@ -3,62 +3,97 @@
  */
 export class CacheService {
-    private _cacheContainer = {};
+    private _cacheContainer:{[key: string]: [number, Promise<any>]} = {};
     private _defaultCacheTtlSeconds = 10;
 
+    /**
+     *
+     * @param key The key for cached value
+     * @param loadingFunction The loading function returning a promise of the value.
+     * @param cacheTtlSeconds The accepted cache timeout in seconds for this call.
+     */
     getForKeyWithLoadingFunction <T> (
         key:string,
         loadingFunction:() => Promise<T>,
         cacheTtlSeconds = this._defaultCacheTtlSeconds
     ):Promise<T> {
-        let cacheEntry = this._cacheContainer[key];
-        const now = (Date.now()/1000);
-        if (cacheEntry !== undefined) {
-            if (cacheEntry[0] >= now - cacheTtlSeconds) {
-                return cacheEntry[1];
-            }
+        const cacheEntry = this._cacheContainer[key];
+        const now = Date.now();
+
+        if (
+            cacheEntry !== undefined
+            && cacheTtlSeconds > 0
+            && cacheEntry[0] >= now
+        ) {
+            return cacheEntry[1];
         }
 
+        const cacheTtlMs = cacheTtlSeconds * 1000;
+        const validUntil = now + cacheTtlMs;
         this._cacheContainer[key] = [
-            now,
-            loadingFunction().then(object=>{
+            validUntil,
+            loadingFunction().then(object => {
                 return this._cacheContainer[key][1] = Promise.resolve(object);
             })
         ];
         return this._cacheContainer[key][1];
     }
 
+    /**
+     * Changes the default cache TTL.
+     * @param seconds
+     */
     setDefaultCacheTtl(seconds:number) {
         this._defaultCacheTtlSeconds = seconds;
         return this;
     }
 
+    /**
+     * Invalidates a list of keys
+     * @param keys
+     */
     invalidate(keys:string[]) {
-        for (let key of keys) {
+        for (const key of keys) {
             delete this._cacheContainer[key];
         }
         return this;
     }
 
+    /**
+     * Returns a list of outdated keys.
+     */
     get outdatedKeys() {
-        const now = (Date.now()/1000);
+        const now = Date.now();
         let cacheEntry;
         const keys = [];
-        for (let key in this._cacheContainer) {
+        for (const key in this._cacheContainer) {
             cacheEntry = this._cacheContainer[key];
-            if (cacheEntry[0] < now - this._defaultCacheTtlSeconds) {
+            if (cacheEntry[0] < now) {
                 keys.push(key);
             }
         }
         return keys;
     }
 
+    /**
+     * A list of all existing cache keys.
+     */
     get cacheKeys() {
         return Object.keys(this._cacheContainer);
     }
 
+    /**
+     * Clears all caches
+     */
     invalidateAll() {
         this._cacheContainer = {};
         return this;
     }
+
+    /**
+     * Invalidates all outdated keys.
+     */
+    invalidateOutdated() {
```

Comment:
No one invalidates the cache keys except the `invalidate` method, which collects "garbage" and frees memory. Thats why you should call that periodically.

Reply:
I've updated the documentation in `README.md` to point that out.

Reply:
As I understand it, that's one of the things that @Zsar wanted to address: Expected behaviour for a cache typically is to invalidate itself (or just a value) upon being read (or so do I believe). Why doesn't this cache do that?

No strong opinions from my side, too little experience.

Reply:
It does invalidate the cache for the particular key (and value) when you call the `get` method.

- Call `getForKeyWithLoadingFunction`
- Key: `foo`
- Loading function: `function () { return Promise.resolve("Katze"); }`

1. When a promise value for the key exists:
2. Check if the promise value is older than the valid TTL.
3. When the promise value is newer
  - Skip calling the loading function
  - Return it immediately
4. Otherwise:
  - Set the value to the loading promise
  - Call the given loading function's promise
  - Write the loading function's promise to the cache
  - Return the loading functions' promise
  
The method does not invalide other keys, when the never get called. Thats why you need to collect the garbage using `invalidateOutdated` to free the memory.

Reply:
Thanks for clarifying, I think I get it now.
So the `invalidateOutdated` function is for clearing the cache of values that are rarely used and take up memory.
Frequently used values invalidate themselves.